### ISSUE TITLE
Include other fobos files

### ISSUE THREAD

**Issue by DaanVanVugt** [szaghi/FoBiS#85]
Hi,

Would it at all be possible to include a file to set some machine-specific library paths?
Then we could have a main fobos file under version control and a fobos.inc file where the user can add some libraries to a template used in the main file.

Thanks,
Daan

**Comment by szaghi** [comment]
@Exteris

Dear Daan,

I think this behavior could be already achieved, but I am not sure. Can you provide a minimum example of what you like to do? Something like what you want to place in main fobos and into children files.

In case your feature cannot be achieved by means of the current FoBis I will be very happy to implement it, however I am now in my holidays I will come back in front of a PC not before 8th August... You have to be patient, I am sorry.

See you soon.

**Comment by Exteris** [comment]
@szaghi 

I would like to have a version-controlled fobos file with compilation profiles etc, and a fobos.inc file with (additions to) some variables, such as paths for specific libraries.
See for example the Makefile structure in a project like SCOTCH: http://gforge.inria.fr/frs/?group_id=248
This has a Make.inc directory with presets for common operating systems, that you need to set some paths in to locations that you have installed libraries in. Many of the presets can be caught by the template system in FoBiS, but machine-specific paths are not convenient I think. (At least I don't know how to do it ;)

Enjoy your holiday!
Daan

**Comment by szaghi** [comment]
@Exteris

I have difficult to see scotch example by means of my smartphone. Anyhow, I think that for your desiderata FoBiS needs some extensions. Global variables can be already set into fobos file, but I think you want to put preset paths into each own fobos.inc file... It should be a very simple extension to FoBiS, but I am not able to do it with my smartphone :smile: when I come back I will do it. See you soon.

**Comment by szaghi** [comment]
Dear @Exteris 

I have now few time to work on this feature. Due to my very limited time, it is very helpful to have some more detailed "instructions" from you. Please, consider the following proposal

#### fobos.inc

```ini
[arch]
lapack = /opt/arch/lapack3.1.1
fft = /opt/arch/fft2.1.0

[ubuntu]
lapack = /opt/ubuntu/lapack3.1.1
fft = /opt/ubuntu/fft2.1.0
```

#### fobos

```ini
[default]
compiler = gnu
libs = $fobos.inc[arch](lapack) $fobos.inc[ubuntu](fft)
```

where

+ `fobos.inc` can be any name of a fobos file;
+ `arch, ubuntu` can be any section name (different from the *rule* sections):
+ `lapack, fft` can be any option name

Does this proposal fit your needs?

**Comment by Exteris** [comment]
Dear @szaghi 

The heart of the matter is the separation of rules for the project and machine-specific rules, detailing things such as which compilers to use (mpifort, mpiifort, whatever these are named), compiler flags and locations of libraries.

Build rules and general configurations should be under git, but machine-specific configurations (and debug options etc) not (in my opinion). For this some templates can be provided, but it is quite inconvenient to have a 'dirty' makefile in your directory, both with regard to accidentally commiting and with changing things in the project-wide makefile.

To make this your approach would be sufficient, but different section names are unnecessary.
I think it would be best to only give a warning if a key does not exist, in that way libraries can be disabled without changing the main fobos file.

Thank you for writing this proposal. Please do not feel hurried, I have implemented a workable GNU Make solution for now.

Best,
Daan

**Comment by szaghi** [comment]
Hi Daan,

> The heart of the matter is the separation of rules for the project and machine-specific rules, detailing things such as which compilers to use (mpifort, mpiifort, whatever these are named), compiler flags and locations of libraries

You have to be patient with, my English is very poor. My understanding was that `fobos` contains the **project rules** and `fobos.inc` contains the **machine specific ones**, is it right?

> Build rules and general configurations should be under git, but machine-specific configurations (and debug options etc) not (in my opinion). For this some templates can be provided, but it is quite inconvenient to have a 'dirty' makefile in your directory, both with regard to accidentally commiting and with changing things in the project-wide makefile.

Yes, we like to version `fobos`, but not `fobos.inc`, is it right?

> To make this your approach would be sufficient, but different section names are unnecessary.

Oh, mine was an overkill... you just want **only 1 container-section** (whatever named), is it right? 

> I think it would be best to only give a warning if a key does not exist, in that way libraries can be disabled without changing the main fobos file.

Yes, this is could be viable. 

To summarize my new understanding you want something like:

#### fobos.inc

```ini
[section-of-any-name]
lapack = /opt/arch/lapack3.1.1
```

#### fobos

```ini
[default]
compiler = gnu
libs = $fobos.inc(lapack) $fobos.inc(fft)
```

that will warn for missing `fft` key without stopping (or reverting back to some default value for the library).

What do you think?

**Comment by Exteris** [comment]
Yes, exactly! I can imagine it being nice to have the different sections
though, so it might be best to keep that in.
Thanks!
